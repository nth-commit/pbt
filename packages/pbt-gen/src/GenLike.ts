import { pipe } from 'ix/iterable';
import { map } from 'ix/iterable/operators';
import { Seed, Size } from 'pbt-core';

export type GenLike<T> = (seed: Seed, size: Size) => Iterable<T>;

export const mapGenLike = <T, U>(g: GenLike<T>, f: (x: T) => U): GenLike<U> => (seed, size) =>
  pipe(g(seed, size), map(f));
