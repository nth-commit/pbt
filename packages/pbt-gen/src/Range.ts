import { Size } from 'pbt-core';

export type Range = {
  getSizedBounds: (size: Size) => [min: number, max: number];
  origin: number;
};

export namespace Range {
  const sort = (x: number, y: number): [min: number, max: number] => {
    const min = x < y ? x : y;
    const max = y > x ? y : x;
    return [min, max];
  };

  const clamp = (min: number, max: number, n: number): number => Math.min(max, Math.max(min, n));

  export const constant = (x: number, y: number): Range => {
    const [min, max] = sort(x, y);

    return {
      getSizedBounds: () => [min, max],
      origin: min,
    };
  };

  export const linear = (x: number, y: number): Range => {
    const [min, max] = sort(x, y);

    return {
      getSizedBounds: (size) => {
        const sizeRatio = size / 100;
        const diff = max - min;
        const scaledMax = Math.round(diff * sizeRatio) + min;
        const clamped = clamp(min, max, scaledMax);
        return [min, clamped];
      },
      origin: min,
    };
  };
}
