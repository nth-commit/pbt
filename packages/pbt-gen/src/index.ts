import { pipe, empty } from 'ix/iterable';
import { map } from 'ix/iterable/operators';
import { Gen as IGen, GenResult, Seed, Size } from 'pbt-core';

export type Gen<T> = IGen<T> & {
  map: <U>(f: (x: T) => U) => Gen<U>;
};

export const create = <T>(g: (seed: Seed, size: Size) => Iterable<T>): Gen<T> => {
  const baseGen: IGen<T> = (seed, size) =>
    map<T, GenResult<T>>((x) => ({
      kind: 'instance',
      value: x,
      shrink: /* istanbul ignore next */ () => empty(),
    }))(g(seed, size));

  const gen: Gen<T> = baseGen as Gen<T>;
  gen.map = null as any;

  return gen;
};

export const integer = (): Gen<number> =>
  create((seed) =>
    pipe(
      Seed.stream(seed),
      map((s) => s.nextInt(0, Number.MAX_SAFE_INTEGER)),
    ),
  );
