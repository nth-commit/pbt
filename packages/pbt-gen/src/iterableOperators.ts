import { OperatorFunction, UnaryFunction } from 'ix/interfaces';
import { IterableX, pipe } from 'ix/iterable';
import { map, tap } from 'ix/iterable/operators';

const getMaxIterations = (): number | undefined => (globalThis as any).__pbtInfiniteStreamProtection;

const crashAfter = <T>(maxIterations: number): UnaryFunction<Iterable<T>, IterableX<T>> => {
  let iterationCount = 0;
  return tap(() => {
    iterationCount++;
    /* istanbul ignore next */
    if (iterationCount > maxIterations) {
      throw new Error('Infinite loop protection');
    }
  });
};

export const addInfiniteStreamProtection = <T>(): OperatorFunction<T, T> => (source: Iterable<T>): IterableX<T> => {
  const maxIterations = getMaxIterations();
  /* istanbul ignore next */
  return maxIterations === undefined ? pipe(source) : pipe(source, crashAfter(maxIterations));
};

export class TakeWhileInclusiveIterable<TSource> extends IterableX<TSource> {
  private _source: Iterable<TSource>;
  private _predicate: (value: TSource, index: number) => boolean;

  constructor(source: Iterable<TSource>, predicate: (value: TSource, index: number) => boolean) {
    super();
    this._source = source;
    this._predicate = predicate;
  }

  *[Symbol.iterator]() {
    let i = 0;
    for (const item of this._source) {
      yield item;
      if (!this._predicate(item, i++)) {
        break;
      }
    }
  }
}

export const takeWhileInclusive = <T>(predicate: (value: T, index: number) => boolean): OperatorFunction<T, T> => (
  source: Iterable<T>,
): IterableX<T> => new TakeWhileInclusiveIterable<T>(source, predicate);

export type Indexed<T> = { value: T; index: number };

export const indexed = <TSource>(): OperatorFunction<TSource, { value: TSource; index: number }> => {
  return map((value, index) => ({ value, index }));
};
