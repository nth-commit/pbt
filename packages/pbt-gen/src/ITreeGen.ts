import { pipe, of, empty } from 'ix/iterable';
import { map, flatMap, share } from 'ix/iterable/operators';
import { GenDiscard, GenExhaustion, GenInstance, Seed, Size } from 'pbt-core';
import { GenLike } from './GenLike';
import { Tree } from './Tree';

export type TreeGenInstance<T> = Omit<GenInstance<Tree<T>>, 'shrink'>;

export type TreeGenResult<T> = TreeGenInstance<T> | GenDiscard | GenExhaustion;

export type ITreeGen<T> = GenLike<TreeGenResult<T>>;

export namespace ITreeGen {
  const bindResult = <T, U>(r: TreeGenResult<T>, k: (x: T) => ITreeGen<U>): ITreeGen<U> => (seed, size) => {
    if (r.kind !== 'instance') return of(r);

    const treeFolder = (outcome0: T, rs: Iterable<TreeGenResult<U>>): Iterable<TreeGenResult<U>> => {
      const treeGenK = k(outcome0);

      return pipe(
        treeGenK(seed, size),
        map(
          (r): TreeGenResult<U> => {
            if (r.kind !== 'instance') return r;

            return { kind: 'exhaustion' };
          },
        ),
      );
    };

    const forestFolder = (rss: Iterable<Iterable<TreeGenResult<U>>>): Iterable<TreeGenResult<U>> =>
      pipe(
        rss,
        flatMap((x) => x),
      );

    return Tree.fold<T, Iterable<TreeGenResult<U>>, Iterable<TreeGenResult<U>>>(r.value, treeFolder, forestFolder);
  };

  export const bind = <T, U>(treeGenBase: ITreeGen<T>, k: (x: T) => ITreeGen<U>): ITreeGen<U> => (seed, size) => {
    const [leftSeed, rightSeed] = seed.split();
    return pipe(
      treeGenBase(leftSeed, size),
      flatMap((r) => bindResult(r, k)(rightSeed, size)),
    );
  };
}
