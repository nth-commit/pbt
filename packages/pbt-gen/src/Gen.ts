import { pipe } from 'ix/iterable';
import { map } from 'ix/iterable/operators';
import { Gen as IGen, GenInstanceData, GenResult, Seed, Size } from 'pbt-core';
import { GenLike } from './GenLike';
import { addInfiniteStreamProtection, takeWhileInclusive } from './iterableOperators';
import { Shrink } from './Shrink';
import { Tree } from './Tree';
import { TreeGen } from './TreeGen';
import { TreeGenResult } from './TreeGenResult';
import * as SeedExtensions from './SeedExtensions';

export type Gen<T> = IGen<T> & {
  map: <U>(f: (x: T) => U) => Gen<U>;
  filter: (f: (x: T) => boolean) => Gen<T>;
  flatMap: <U>(k: (x: T) => Gen<U>) => Gen<U>;
  noShrink: () => Gen<T>;
};

const mapTreeToInstanceData = <T>([outcome, shrinks]: Tree<T>): GenInstanceData<T> => ({
  value: outcome,
  shrink: () => pipe(shrinks, map(mapTreeToInstanceData)),
});

const mapTreeGenResultToGenResult = <T>(r: TreeGenResult<T>): GenResult<T> =>
  r.kind === 'instance'
    ? {
        kind: 'instance',
        ...mapTreeToInstanceData(r.value),
      }
    : r;

const mapTreeGenToBaseGen = <T>(gTree: TreeGen<T>): IGen<T> => (seed, size) =>
  pipe(
    gTree(seed, size),
    takeWhileInclusive((r) => r.kind !== 'exhaustion'),
    addInfiniteStreamProtection(),
    map(mapTreeGenResultToGenResult),
  );

const mapGenKToTreeGenK = <T, U>(k: (x: T) => Gen<U>) => (x: T): TreeGen<U> => TreeGen.fromGen<U>(k(x));

const mapTreeGenToGen = <T>(gTree: TreeGen<T>): Gen<T> => {
  const genMap = <U>(f: (x: T) => U): Gen<U> => mapTreeGenToGen<U>(gTree.map(f));

  const genFilter = (f: (x: T) => boolean): Gen<T> => mapTreeGenToGen(gTree.filter(f));

  const genFlatMap = <U>(k: (x: T) => Gen<U>): Gen<U> => mapTreeGenToGen(gTree.flatMap(mapGenKToTreeGenK(k)));

  const genNoShrink = (): Gen<T> => mapTreeGenToGen(gTree.noShrink());

  return Object.assign(mapTreeGenToBaseGen(gTree), {
    map: genMap,
    filter: genFilter,
    flatMap: genFlatMap,
    noShrink: genNoShrink,
  });
};

export const stream = <T>(g: GenLike<T>, shrink: Shrink<T>): Gen<T> => {
  const gTree = TreeGen.fromGenLike(g, shrink);
  return mapTreeGenToGen(gTree);
};

export const create = <T>(f: (seed: Seed, size: Size) => T, shrink: Shrink<T>): Gen<T> =>
  stream(
    (seed, size) =>
      pipe(
        SeedExtensions.stream(seed),
        map((seed0) => f(seed0, size)),
      ),
    shrink,
  );

export const exhausted = <T>(): Gen<T> => {
  const gTree = TreeGen.exhausted<T>();
  return mapTreeGenToGen(gTree);
};
