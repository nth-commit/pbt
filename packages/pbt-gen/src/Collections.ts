import { Gen, create } from './Gen';
import { Range } from './Range';
import { Shrink } from './Shrink';

const arrayFromRange = <T>(range: Range, gElement: Gen<T>): Gen<T[]> =>
  create((seed, size) => {
    const [leftSeed, rightSeed] = seed.split();
    const [minLength, maxLength] = range.getSizedBounds(size);
    const length = leftSeed.nextInt(minLength, maxLength);

    const g = gElement.reduce<T[]>(length, (arr, x) => [...arr, x], []).preShrink(Shrink.array(minLength));

    return g(rightSeed, size);
  });

export const array = {
  unscaled: <T>(min: number, max: number, g: Gen<T>): Gen<T[]> => arrayFromRange(Range.constant(min, max), g),

  scaleLinearly: <T>(min: number, max: number, g: Gen<T>): Gen<T[]> => arrayFromRange(Range.linear(min, max), g),
};
