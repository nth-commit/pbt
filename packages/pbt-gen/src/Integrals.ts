import { pipe } from 'ix/iterable';
import { map } from 'ix/iterable/operators';
import { Seed, Size } from 'pbt-core';
import { create, Gen } from './Gen';
import { Shrink } from './Shrink';

type Range = {
  getSizedBounds: (size: Size) => [min: number, max: number, maxDp: number];
  origin: number;
};

namespace Range {
  const sort = (x: number, y: number): [min: number, max: number] => {
    const min = x < y ? x : y;
    const max = y > x ? y : x;
    return [min, max];
  };

  const clamp = (min: number, max: number, n: number): number => Math.min(max, Math.max(min, n));

  export const constant = (x: number, y: number, dp: number): Range => {
    const [min, max] = sort(x, y);

    return {
      getSizedBounds: () => [min, max, dp],
      origin: min,
    };
  };

  export const linear = (x: number, y: number, dp: number): Range => {
    const [min, max] = sort(x, y);

    return {
      getSizedBounds: (size) => {
        const sizeRatio = size / 100;
        const diff = max - min;
        const scaledMax = Math.round(diff * sizeRatio) + min;
        const clamped = clamp(min, max, scaledMax);
        return [min, clamped, dp];
      },
      origin: min,
    };
  };
}

const nextNumber = (size: Size, range: Range) => (seed: Seed): number => {
  const [min, max, maxDp] = range.getSizedBounds(size);

  /* istanbul ignore next */
  if (min > max) throw new Error('This causes the random to hang...');

  /* istanbul ignore next */
  if (maxDp === 0) {
    return seed.nextInt(min, max);
  }

  /* istanbul ignore next */
  throw new Error(`Unsupported: Number of decimal places ${maxDp}`);
};

const integral = (range: Range): Gen<number> =>
  create((seed, size) => pipe(Seed.stream(seed), map(nextNumber(size, range))), Shrink.towardsNumber(range.origin, 0));

export const integer = {
  constant: (min: number, max: number): Gen<number> => integral(Range.constant(min, max, 0)),

  linear: (min: number, max: number): Gen<number> => integral(Range.linear(min, max, 0)),
};
