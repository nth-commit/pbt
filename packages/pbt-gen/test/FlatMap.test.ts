import fc from 'fast-check';
import { toArray, pipe } from 'ix/iterable';
import { map, take } from 'ix/iterable/operators';
import * as devCore from 'pbt-core';
import * as dev from '../src';
import * as stable from './helpers/stableApi';
import {
  arbitraryGenParams,
  arbitraryIterations,
  arbitraryGenerator,
  arbitraryFunction,
  arbitraryFullGenerator,
} from './helpers/arbitraries';
import { castToInstance } from './helpers/iterableOperators';
import { GenInstance } from 'pbt-core';

const arbitraryGenFlatMapper = <T>(): fc.Arbitrary<(x: T) => dev.Gen<unknown>> =>
  arbitraryFunction<dev.Gen<unknown>>(arbitraryGenerator(), 1);

test('It exhausts if the left generator exhausts', () => {
  stable.assert(
    stable.property(arbitraryGenParams(), arbitraryGenFlatMapper(), ({ seed, size }, k) => {
      const gLeft = dev.exhausted();
      const g = gLeft.flatMap(k);

      const results = toArray(pipe(g(seed, size)));

      expect(results).toEqual([{ kind: 'exhaustion' }]);
    }),
  );
});

test('It exhausts if the right generator exhausts', () => {
  stable.assert(
    stable.property(arbitraryGenParams(), arbitraryFullGenerator(), ({ seed, size }, gLeft) => {
      const g = gLeft.flatMap(() => dev.exhausted());

      const results = toArray(pipe(g(seed, size), take(3)));

      expect(results).toEqual([{ kind: 'exhaustion' }]);
    }),
  );
});

test('It discards when the left generator discards', () => {
  stable.assert(
    stable.property(
      arbitraryGenParams(),
      arbitraryIterations(),
      arbitraryGenerator(),
      arbitraryGenFlatMapper(),
      ({ seed, size }, iterations, gLeft, k) => {
        const g = gLeft.filter(() => false).flatMap(k);

        const results = toArray(pipe(g(seed, size), take(iterations)));

        expect(results).toHaveLength(iterations);
        results.forEach((result) => {
          expect(result).toEqual({ kind: 'discard' });
        });
      },
    ),
  );
});

test('It discards when the right generator discards', () => {
  stable.assert(
    stable.property(
      arbitraryGenParams(),
      arbitraryIterations(),
      arbitraryGenerator(),
      arbitraryGenerator(),
      ({ seed, size }, iterations, gLeft, gRight) => {
        const g = gLeft.flatMap(() => gRight.filter(() => false));

        const results = toArray(pipe(g(seed, size), take(iterations)));

        expect(results).toHaveLength(iterations);
        results.forEach((result) => {
          expect(result).toEqual({ kind: 'discard' });
        });
      },
    ),
  );
});

test('Snapshot', () => {
  // It's hard to describe the intricacies of how flatMap combines streams. Let's just generate examples for now rather
  // than sinking a huge amount of time into it.

  const seed = devCore.Seed.create(0);
  const iterationsBySize = new Map<number, number>([
    [0, 1],
    [20, 2],
    [50, 5],
    [100, 1],
  ]);
  const g = dev.integer.linear(0, 5).flatMap((x) => dev.integer.linear(0, 5).map((y) => `[${x},${y}]`));

  for (const [size, iterations] of iterationsBySize.entries()) {
    const results = toArray(pipe(g(seed, size), castToInstance(), map(GenInstance.evaluate), take(iterations)));

    results.forEach((evaluatedInstance, i) =>
      expect(JSON.stringify(evaluatedInstance, null, 2).replace(/"/g, '')).toMatchSnapshot(
        `size=${size} iteration=${i + 1}`,
      ),
    );
  }
});
