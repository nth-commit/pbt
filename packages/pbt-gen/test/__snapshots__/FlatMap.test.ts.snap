// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: size=0 iteration=1 1`] = `"[0,0]"`;

exports[`Snapshot: size=20 iteration=1 1`] = `
"[
  [0,1],
  [
    [0,0]
  ]
]"
`;

exports[`Snapshot: size=20 iteration=2 1`] = `
"[
  [1,0],
  [
    [0,0]
  ]
]"
`;

exports[`Snapshot: size=50 iteration=1 1`] = `
"[
  [0,1],
  [
    [0,0]
  ]
]"
`;

exports[`Snapshot: size=50 iteration=2 1`] = `
"[
  [1,2],
  [
    [
      [0,2],
      [
        [0,0],
        [
          [0,1],
          [
            [0,0]
          ]
        ]
      ]
    ],
    [1,0],
    [
      [1,1],
      [
        [1,0]
      ]
    ]
  ]
]"
`;

exports[`Snapshot: size=50 iteration=3 1`] = `
"[
  [0,2],
  [
    [0,0],
    [
      [0,1],
      [
        [0,0]
      ]
    ]
  ]
]"
`;

exports[`Snapshot: size=50 iteration=4 1`] = `
"[
  [3,2],
  [
    [
      [0,2],
      [
        [0,0],
        [
          [0,1],
          [
            [0,0]
          ]
        ]
      ]
    ],
    [
      [1,2],
      [
        [
          [0,2],
          [
            [0,0],
            [
              [0,1],
              [
                [0,0]
              ]
            ]
          ]
        ],
        [1,0],
        [
          [1,1],
          [
            [1,0]
          ]
        ]
      ]
    ],
    [
      [2,2],
      [
        [
          [0,2],
          [
            [0,0],
            [
              [0,1],
              [
                [0,0]
              ]
            ]
          ]
        ],
        [
          [1,2],
          [
            [
              [0,2],
              [
                [0,0],
                [
                  [0,1],
                  [
                    [0,0]
                  ]
                ]
              ]
            ],
            [1,0],
            [
              [1,1],
              [
                [1,0]
              ]
            ]
          ]
        ],
        [2,0],
        [
          [2,1],
          [
            [2,0]
          ]
        ]
      ]
    ],
    [3,0],
    [
      [3,1],
      [
        [3,0]
      ]
    ]
  ]
]"
`;

exports[`Snapshot: size=50 iteration=5 1`] = `
"[
  [0,3],
  [
    [0,0],
    [
      [0,1],
      [
        [0,0]
      ]
    ],
    [
      [0,2],
      [
        [0,0],
        [
          [0,1],
          [
            [0,0]
          ]
        ]
      ]
    ]
  ]
]"
`;

exports[`Snapshot: size=100 iteration=1 1`] = `
"[
  [4,5],
  [
    [
      [0,5],
      [
        [0,0],
        [
          [0,2],
          [
            [0,0],
            [
              [0,1],
              [
                [0,0]
              ]
            ]
          ]
        ],
        [
          [0,3],
          [
            [0,0],
            [
              [0,1],
              [
                [0,0]
              ]
            ],
            [
              [0,2],
              [
                [0,0],
                [
                  [0,1],
                  [
                    [0,0]
                  ]
                ]
              ]
            ]
          ]
        ],
        [
          [0,4],
          [
            [0,0],
            [
              [0,2],
              [
                [0,0],
                [
                  [0,1],
                  [
                    [0,0]
                  ]
                ]
              ]
            ],
            [
              [0,3],
              [
                [0,0],
                [
                  [0,1],
                  [
                    [0,0]
                  ]
                ],
                [
                  [0,2],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ]
      ]
    ],
    [
      [2,5],
      [
        [
          [0,5],
          [
            [0,0],
            [
              [0,2],
              [
                [0,0],
                [
                  [0,1],
                  [
                    [0,0]
                  ]
                ]
              ]
            ],
            [
              [0,3],
              [
                [0,0],
                [
                  [0,1],
                  [
                    [0,0]
                  ]
                ],
                [
                  [0,2],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [
              [0,4],
              [
                [0,0],
                [
                  [0,2],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ]
                  ]
                ],
                [
                  [0,3],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ],
        [
          [1,5],
          [
            [
              [0,5],
              [
                [0,0],
                [
                  [0,2],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ]
                  ]
                ],
                [
                  [0,3],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ],
                [
                  [0,4],
                  [
                    [0,0],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ],
                    [
                      [0,3],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ],
                        [
                          [0,2],
                          [
                            [0,0],
                            [
                              [0,1],
                              [
                                [0,0]
                              ]
                            ]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [1,0],
            [
              [1,2],
              [
                [1,0],
                [
                  [1,1],
                  [
                    [1,0]
                  ]
                ]
              ]
            ],
            [
              [1,3],
              [
                [1,0],
                [
                  [1,1],
                  [
                    [1,0]
                  ]
                ],
                [
                  [1,2],
                  [
                    [1,0],
                    [
                      [1,1],
                      [
                        [1,0]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [
              [1,4],
              [
                [1,0],
                [
                  [1,2],
                  [
                    [1,0],
                    [
                      [1,1],
                      [
                        [1,0]
                      ]
                    ]
                  ]
                ],
                [
                  [1,3],
                  [
                    [1,0],
                    [
                      [1,1],
                      [
                        [1,0]
                      ]
                    ],
                    [
                      [1,2],
                      [
                        [1,0],
                        [
                          [1,1],
                          [
                            [1,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ],
        [2,0],
        [
          [2,2],
          [
            [2,0],
            [
              [2,1],
              [
                [2,0]
              ]
            ]
          ]
        ],
        [
          [2,3],
          [
            [2,0],
            [
              [2,1],
              [
                [2,0]
              ]
            ],
            [
              [2,2],
              [
                [2,0],
                [
                  [2,1],
                  [
                    [2,0]
                  ]
                ]
              ]
            ]
          ]
        ],
        [
          [2,4],
          [
            [2,0],
            [
              [2,2],
              [
                [2,0],
                [
                  [2,1],
                  [
                    [2,0]
                  ]
                ]
              ]
            ],
            [
              [2,3],
              [
                [2,0],
                [
                  [2,1],
                  [
                    [2,0]
                  ]
                ],
                [
                  [2,2],
                  [
                    [2,0],
                    [
                      [2,1],
                      [
                        [2,0]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ]
      ]
    ],
    [
      [3,5],
      [
        [
          [0,5],
          [
            [0,0],
            [
              [0,2],
              [
                [0,0],
                [
                  [0,1],
                  [
                    [0,0]
                  ]
                ]
              ]
            ],
            [
              [0,3],
              [
                [0,0],
                [
                  [0,1],
                  [
                    [0,0]
                  ]
                ],
                [
                  [0,2],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [
              [0,4],
              [
                [0,0],
                [
                  [0,2],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ]
                  ]
                ],
                [
                  [0,3],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ],
        [
          [1,5],
          [
            [
              [0,5],
              [
                [0,0],
                [
                  [0,2],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ]
                  ]
                ],
                [
                  [0,3],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ],
                [
                  [0,4],
                  [
                    [0,0],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ],
                    [
                      [0,3],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ],
                        [
                          [0,2],
                          [
                            [0,0],
                            [
                              [0,1],
                              [
                                [0,0]
                              ]
                            ]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [1,0],
            [
              [1,2],
              [
                [1,0],
                [
                  [1,1],
                  [
                    [1,0]
                  ]
                ]
              ]
            ],
            [
              [1,3],
              [
                [1,0],
                [
                  [1,1],
                  [
                    [1,0]
                  ]
                ],
                [
                  [1,2],
                  [
                    [1,0],
                    [
                      [1,1],
                      [
                        [1,0]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [
              [1,4],
              [
                [1,0],
                [
                  [1,2],
                  [
                    [1,0],
                    [
                      [1,1],
                      [
                        [1,0]
                      ]
                    ]
                  ]
                ],
                [
                  [1,3],
                  [
                    [1,0],
                    [
                      [1,1],
                      [
                        [1,0]
                      ]
                    ],
                    [
                      [1,2],
                      [
                        [1,0],
                        [
                          [1,1],
                          [
                            [1,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ],
        [
          [2,5],
          [
            [
              [0,5],
              [
                [0,0],
                [
                  [0,2],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ]
                  ]
                ],
                [
                  [0,3],
                  [
                    [0,0],
                    [
                      [0,1],
                      [
                        [0,0]
                      ]
                    ],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ],
                [
                  [0,4],
                  [
                    [0,0],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ],
                    [
                      [0,3],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ],
                        [
                          [0,2],
                          [
                            [0,0],
                            [
                              [0,1],
                              [
                                [0,0]
                              ]
                            ]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [
              [1,5],
              [
                [
                  [0,5],
                  [
                    [0,0],
                    [
                      [0,2],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ]
                      ]
                    ],
                    [
                      [0,3],
                      [
                        [0,0],
                        [
                          [0,1],
                          [
                            [0,0]
                          ]
                        ],
                        [
                          [0,2],
                          [
                            [0,0],
                            [
                              [0,1],
                              [
                                [0,0]
                              ]
                            ]
                          ]
                        ]
                      ]
                    ],
                    [
                      [0,4],
                      [
                        [0,0],
                        [
                          [0,2],
                          [
                            [0,0],
                            [
                              [0,1],
                              [
                                [0,0]
                              ]
                            ]
                          ]
                        ],
                        [
                          [0,3],
                          [
                            [0,0],
                            [
                              [0,1],
                              [
                                [0,0]
                              ]
                            ],
                            [
                              [0,2],
                              [
                                [0,0],
                                [
                                  [0,1],
                                  [
                                    [0,0]
                                  ]
                                ]
                              ]
                            ]
                          ]
                        ]
                      ]
                    ]
                  ]
                ],
                [1,0],
                [
                  [1,2],
                  [
                    [1,0],
                    [
                      [1,1],
                      [
                        [1,0]
                      ]
                    ]
                  ]
                ],
                [
                  [1,3],
                  [
                    [1,0],
                    [
                      [1,1],
                      [
                        [1,0]
                      ]
                    ],
                    [
                      [1,2],
                      [
                        [1,0],
                        [
                          [1,1],
                          [
                            [1,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ],
                [
                  [1,4],
                  [
                    [1,0],
                    [
                      [1,2],
                      [
                        [1,0],
                        [
                          [1,1],
                          [
                            [1,0]
                          ]
                        ]
                      ]
                    ],
                    [
                      [1,3],
                      [
                        [1,0],
                        [
                          [1,1],
                          [
                            [1,0]
                          ]
                        ],
                        [
                          [1,2],
                          [
                            [1,0],
                            [
                              [1,1],
                              [
                                [1,0]
                              ]
                            ]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [2,0],
            [
              [2,2],
              [
                [2,0],
                [
                  [2,1],
                  [
                    [2,0]
                  ]
                ]
              ]
            ],
            [
              [2,3],
              [
                [2,0],
                [
                  [2,1],
                  [
                    [2,0]
                  ]
                ],
                [
                  [2,2],
                  [
                    [2,0],
                    [
                      [2,1],
                      [
                        [2,0]
                      ]
                    ]
                  ]
                ]
              ]
            ],
            [
              [2,4],
              [
                [2,0],
                [
                  [2,2],
                  [
                    [2,0],
                    [
                      [2,1],
                      [
                        [2,0]
                      ]
                    ]
                  ]
                ],
                [
                  [2,3],
                  [
                    [2,0],
                    [
                      [2,1],
                      [
                        [2,0]
                      ]
                    ],
                    [
                      [2,2],
                      [
                        [2,0],
                        [
                          [2,1],
                          [
                            [2,0]
                          ]
                        ]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ],
        [3,0],
        [
          [3,2],
          [
            [3,0],
            [
              [3,1],
              [
                [3,0]
              ]
            ]
          ]
        ],
        [
          [3,3],
          [
            [3,0],
            [
              [3,1],
              [
                [3,0]
              ]
            ],
            [
              [3,2],
              [
                [3,0],
                [
                  [3,1],
                  [
                    [3,0]
                  ]
                ]
              ]
            ]
          ]
        ],
        [
          [3,4],
          [
            [3,0],
            [
              [3,2],
              [
                [3,0],
                [
                  [3,1],
                  [
                    [3,0]
                  ]
                ]
              ]
            ],
            [
              [3,3],
              [
                [3,0],
                [
                  [3,1],
                  [
                    [3,0]
                  ]
                ],
                [
                  [3,2],
                  [
                    [3,0],
                    [
                      [3,1],
                      [
                        [3,0]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ]
      ]
    ],
    [4,0],
    [
      [4,2],
      [
        [4,0],
        [
          [4,1],
          [
            [4,0]
          ]
        ]
      ]
    ],
    [
      [4,3],
      [
        [4,0],
        [
          [4,1],
          [
            [4,0]
          ]
        ],
        [
          [4,2],
          [
            [4,0],
            [
              [4,1],
              [
                [4,0]
              ]
            ]
          ]
        ]
      ]
    ],
    [
      [4,4],
      [
        [4,0],
        [
          [4,2],
          [
            [4,0],
            [
              [4,1],
              [
                [4,0]
              ]
            ]
          ]
        ],
        [
          [4,3],
          [
            [4,0],
            [
              [4,1],
              [
                [4,0]
              ]
            ],
            [
              [4,2],
              [
                [4,0],
                [
                  [4,1],
                  [
                    [4,0]
                  ]
                ]
              ]
            ]
          ]
        ]
      ]
    ]
  ]
]"
`;
